### DISCUSSION TITLE
What's the best way to login in manually.

### DISCUSSION THREAD

**Discussion by VincentLanglet** [symfony/symfony#48964] [Q&A] [answered]
I'm using Symfony 5.4 and when I want to login a user manually, I'm currently doing:
```
$token = new UsernamePasswordToken($user, null, 'main_context', $user->getRoles());
$this->tokenStorage->setToken($token);
$this->session->set('_security_main', serialize($token));
```

With all the changes done to authentication, I wonder if there is not something provided by Symfony to do this.
I first thought about `UserAuthenticatorInterface::authenticateUser`, but this require both a `AuthenticatorInterface` and a `Request` which I doesn't need in my snippet...

**Comment by stof** [comment]
Your snippet works only for specific cases (not all authenticators rely on a UsernamePasswordToken). `UserAuthenticatorInterface::authenticateUser` taking the authenticator is meant to support any kind of authenticator (so that the right kind of token is created).

And your snippet technically requires a Request at some earlier point to access the Session. It is just not shown in the current snippet.

**Reply by VincentLanglet** [reply]
Didn't think about request because I'm still injecting the SessionInterface ; but indeed I'll need to change this for Sf6.

Then, if I want to keep a similar behavior with the new syntax, I assume, I'll need something like
```
$userAuthenticator->authenticateUser($user, ???, $requestStack->getRequest());
```
But which Authenticator should I pass ?

**Comment by mbabker** [comment, marked as answer]
Symfony 6.2 has an API to help with [programatic login](https://symfony.com/doc/current/security.html#login-programmatically), use that when able.

**Reply by VincentLanglet** [reply]
Thanks ! I'll try to bump to Symfony 6 before changing this code then

**Reply by stof** [reply]
Indeed, in 6.2+, SecurityBundle exposes a high-level helper on top of the UserAuthenticatorInterface, which is the easiest way
